type_inference/type_retrieval_service.py

```diff
@@ -0,0 +1,139 @@
+    child_type.typname AS field_type
+FROM pg_type AS parent_type
+    JOIN pg_attribute ON pg_attribute.attrelid = parent_type.typrelid
+    JOIN pg_type AS child_type ON child_type.oid = pg_attribute.atttypid
+WHERE parent_type.typname = '{%s}';''', (udt_type,))
+
+    fields = (f'{field_name}: {unpack_type(field_type, conn)}' for field_name, field_type in cur.fetchall())
+    return f'{{{", ".join(fields)}}}'
+
+
+def ValidateRuleAndGetTableName(rule: dict) -> str:
+  rule_text = rule['full_text']
+  field_value = rule['head']['record']['field_value']
+
+  if len(field_value) != 1 or field_value[0]['field'] != '*':
+    raise TypeRetrievalException(rule_text)
+
+  conjuncts = rule['body']['conjunction']['conjunct']
+
+  if len(conjuncts) != 1:
+    raise TypeRetrievalException(rule_text)
+
+  conjunct = conjuncts[0]
+
+  if 'predicate' not in conjunct:
+    raise TypeRetrievalException(rule_text)
+
+  field_values = conjunct['predicate']['record']['field_value']
+
+  if len(field_values) != 1 or field_values[0]['field'] != '*':
+    raise TypeRetrievalException(rule_text)
+
+  return conjuncts[0]['predicate']['predicate_name'].split('.')[1]
+
+
+class TypeRetrievalService:
+  def __init__(self, parsed_rules, predicate_names,
+               connection_string='dbname=logica user=logica password=logica host=127.0.0.1'):
+    predicate_names_as_set = set(predicate_names)
+    self.parsed_rules = [r for r in parsed_rules if r['head']['predicate_name'] in predicate_names_as_set]
+    self.connection_string = connection_string
+    self.table_names = self.ValidateParsedRulesAndGetTableNames()
+    InitBuiltInTypes(self.connection_string)
+
+  def ValidateParsedRulesAndGetTableNames(self) -> Dict[str, str]:
+    return {rule['head']['predicate_name']: ValidateRuleAndGetTableName(rule) for rule in self.parsed_rules}
+
+  def RetrieveTypes(self, filename='default.l'):
+    filename = filename.replace('.l', '_schema.l')
+    with psycopg2.connect(self.connection_string) as conn:
+      with conn.cursor() as cursor:
+        # for each given table this SQL query returns json object
+        # where keys are names of columns in that table and values are corresponding types
+        cursor.execute('''
+SELECT table_name, jsonb_object_agg(column_name, udt_name)
+FROM information_schema.columns
+GROUP BY table_name
+HAVING table_name IN %s;''', (self.table_names.values(),))
+        columns = {table: columns for table, columns in cursor.fetchall()}
+
+      result = []
+
+      for rule in self.parsed_rules:
+        result.append(f'{rule["full_text"]},')
+        local = []
+
+        for column, udt_type in sorted(columns[self.table_names[rule['head']['predicate_name']]].items(), key=lambda t: t[0]):
+          local.append(f'{column}: {unpack_type(udt_type, conn)}')
+
+        var_name = rule['head']['record']['field_value'][0]['value']['expression']['variable']['var_name']
+        result.append(f'{var_name} ~ {{{", ".join(local)}}};{linesep}')
```

Comment:
Please simply use \n rather than linesep.

Reply:
Escaping {} makes code hard to read. It's better to write `'%s ~ {%s};\n' % (var_name, fields)`.
And assign fields to the join with comma.